cpp/src/arrow/compute/util_internal.h

```diff
@@ -26,8 +26,10 @@ namespace arrow {
 namespace util {
 
 template <typename T>
-void CheckAlignment(const void* ptr) {
-  ARROW_DCHECK(reinterpret_cast<uint64_t>(ptr) % sizeof(T) == 0);
+void CheckAlignment(const void* ptr, bool do_check = true) {
```

Comment:
If the columns of the RowTable are not sorted during comparing(that is `are_cols_sorted=false`), then there is no need to check memory alignment.